src/PhpSpec/Console/Application.php

```diff
@@ -57,9 +58,17 @@ public function getContainer()
      */
     public function doRun(InputInterface $input, OutputInterface $output)
     {
+        $helperSet = $this->getHelperSet();
         $this->container->set('console.input', $input);
         $this->container->set('console.output', $output);
-        $this->container->set('console.helper.dialog', $this->getHelperSet()->get('dialog'));
+        $this->container->set('console.helper.dialog', $helperSet->get('dialog'));
```

Comment:
but you already broke BC for this case, as he needs to provide the Prompter to the parent.

For composition, this could be handled by providing a nice way to handle extending a service by composition in the container instead. Both Pimple (since 1.0) and Symfony DIC (since 2.5) have a nice way for this (the Symfony DIC allows to do it since 2.0, but in a less simple way). This would avoid the decorator to know anything about the instantiation of the inner service. This is an issue with the container only if they have to redo all the work.

for inheritance, this would indeed forbid to consider the constructor as an internal API. We don't face this issue in Behat though, because most our classes are final.